Bug: Race condition on stream shutdown leads to "AssertionError: cannot call reset() more than once"

# Bug: Race condition on stream shutdown leads to "AssertionError: cannot call reset() more than once"

### Description

A race condition during the stream shutdown sequence can lead to an unhandled **`AssertionError`** when building higher-level protocols (e.g., WebTransport) on top of `aioquic`.

The issue is triggered when two stream termination events are processed in rapid succession for the same stream:

1. **Application-Driven Reset**: The higher-level protocol logic decides to terminate a stream based on its own semantics. For instance, upon receiving a `CLOSE_WEBTRANSPORT_SESSION` message, our application iterates through associated streams and calls `quic.reset_stream()`.

2. **Transport-Driven Reset**: Simultaneously, the peer, as part of its own shutdown, sends a `STOP_SENDING` frame for the same stream.

If the application's call to `reset_stream()` executes first, the stream's sender is marked as reset. When `aioquic` subsequently processes the incoming `STOP_SENDING` frame, its internal `_handle_stop_sending_frame` handler unconditionally calls `stream.sender.reset()` a second time. This triggers the `AssertionError: cannot call reset() more than once`.

The root cause is that the `_handle_stop_sending_frame` handler is not **idempotent**.

### Proposed Solution

The handler for `STOP_SENDING` frames should be made idempotent by checking if the stream's sender has already been reset before attempting to reset it again. This is a minimal, low-risk change that makes the shutdown process more robust against these predictable race conditions.

- **File**: `aioquic/quic/connection.py`
- **Method**: `_handle_stop_sending_frame`

**Suggested Change**:

```python
# Before
stream = self._get_or_create_stream(frame_type, stream_id)
stream.sender.reset(error_code=QuicErrorCode.NO_ERROR)

# After
stream = self._get_or_create_stream(frame_type, stream_id)
if stream.sender._reset_error_code is None:
    stream.sender.reset(error_code=QuicErrorCode.NO_ERROR)
```

### Workarounds Considered

We have implemented a workaround in our application layer by suppressing this specific AssertionError. This is a suboptimal solution because:

- It forces the application layer to handle a race condition that is more appropriately and robustly handled within the transport layer itself.

- Our workaround must be implemented at a very low level (by subclassing QuicConnectionProtocol) which is intrusive and architecturally undesirable.

- Catching AssertionError is generally an anti-pattern, only justified here as a temporary insulation from an upstream issue.

### Context and Reproduction

This issue was consistently reproduced while developing PyWebTransport, a Python WebTransport library built on aioquic. The error occurs reliably during the teardown phase of our long-running stability and connection churn tests.

### Minimal Reproducible Example Concept

A reliable test would involve a client and server where the client, upon deciding to close, performs two actions concurrently:

- Sends an application-level "close" message on a control stream.

- Immediately closes the QUIC connection, which implicitly queues a STOP_SENDING frame for any active streams.

The server, upon receiving the "close" message, would immediately call reset_stream() on the associated data stream. If this executes before the STOP_SENDING frame is processed from the datagram buffer, the race is triggered.

### Traceback

```
ERROR - Exception in callback _SelectorDatagramTransport._read_ready()
handle: <Handle _SelectorDatagramTransport._read_ready()>
Traceback (most recent call last):
  File "/path/to/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/path/to/python3.12/asyncio/selector_events.py", line 1251, in _read_ready
    self._protocol.datagram_received(data, addr)
  File "/path/to/aioquic/asyncio/server.py", line 149, in datagram_received
    protocol.datagram_received(data, addr)
  File "/path/to/aioquic/asyncio/protocol.py", line 156, in datagram_received
    self._quic.receive_datagram(cast(bytes, data), addr, now=self._loop.time())
  File "/path/to/aioquic/quic/connection.py", line 1032, in receive_datagram
    is_ack_eliciting, is_probing = self._payload_received(...)
  File "/path/to/aioquic/quic/connection.py", line 2453, in _payload_received
    frame_handler(context, frame_type, buf)
  File "/path/to/aioquic/quic/connection.py", line 2206, in _handle_stop_sending_frame
    stream.sender.reset(error_code=QuicErrorCode.NO_ERROR)
  File "/path/to/aioquic/quic/stream.py", line 319, in reset
    assert self._reset_error_code is None, "cannot call reset() more than once"
AssertionError: cannot call reset() more than once
```

Making this handler robust would improve aioquic's stability for complex, stateful applications built on top of it. Thank you for your work on this excellent library.

--- a/src/aioquic/quic/connection.py
+++ b/src/aioquic/quic/connection.py
@@ -1095,6 +1095,9 @@ class QuicConnection:
         """
         Abruptly terminate the sending part of a stream.
 
+        This method has no effect if a reset has already been triggered either by a
+        call to :meth:`reset_stream` or by the reception of a STOP_SENDING frame.
+
         .. aioquic_transmit::
 
         :param stream_id: The stream's ID.

--- a/src/aioquic/quic/stream.py
+++ b/src/aioquic/quic/stream.py
@@ -315,13 +315,16 @@ class QuicStreamSender:
     def reset(self, error_code: int) -> None:
         """
         Abruptly terminate the sending part of the QUIC stream.
-        """
-        assert self._reset_error_code is None, "cannot call reset() more than once"
-        self._reset_error_code = error_code
-        self.reset_pending = True
 
-        # Prevent any more data from being sent or re-sent.
-        self.buffer_is_empty = True
+        Once this method has been called, any further calls to it
+        will have no effect.
+        """
+        if self._reset_error_code is None:
+            self._reset_error_code = error_code
+            self.reset_pending = True
+
+            # Prevent any more data from being sent or re-sent.
+            self.buffer_is_empty = True
 
     def write(self, data: bytes, end_stream: bool = False) -> None:
         """

--- a/tests/test_connection.py
+++ b/tests/test_connection.py
@@ -2291,6 +2291,20 @@ class QuicConnectionTest(TestCase):
 
             self.assertIsNone(client.next_event())
 
+            # client receives another STOP_SENDING
+            client._handle_stop_sending_frame(
+                client_receive_context(client),
+                QuicFrameType.STOP_SENDING,
+                Buffer(data=b"\x00\x12"),
+            )
+
+            event = client.next_event()
+            self.assertEqual(type(event), events.StopSendingReceived)
+            self.assertEqual(event.stream_id, 0)
+            self.assertEqual(event.error_code, 0x12)
+
+            self.assertIsNone(client.next_event())
+
     def test_handle_stop_sending_frame_receive_only(self):
         with client_and_server() as (client, server):
             # server creates unidirectional stream 3

